pandas/core/index.py

```diff
@@ -3818,6 +3838,530 @@ def _wrap_joined_index(self, joined, other):
 Int64Index._add_logical_methods()
 
 
+class RangeIndex(Int64Index):
+
+    """
+    Immutable Index implementing an monotonic range. RangeIndex is a
+    memory-saving special case of Int64Index limited to representing
+    monotonic ranges.
+
+    Parameters
+    ----------
+    start : int (default: 0)
+    stop : int (default: 0)
+    step : int (default: 1)
+    name : object, optional
+        Name to be stored in the index
+    copy : bool, default False
+        Make a copy of input if its a RangeIndex
+
+    Notes
+    -----
+    start can also be a RangeIndex, or a range object
+
+    """
+
+    _typ = 'rangeindex'
+    _engine_type = _index.Int64Engine
+
+    def __new__(cls, start=None, stop=None, step=None, name=None, dtype=None, fastpath=False, copy=False, **kwargs):
+        if fastpath:
+            return cls._simple_new(start, stop, step, name=name)
+
+        # require dtype to be None, int64, or object (for compat)
+        if not (dtype is None or is_int64_dtype(dtype) or is_object_dtype(dtype)):
+            raise TypeError('Invalid to pass a non-int64 dtype to RangeIndex')
+
+        # RangeIndex
+        if isinstance(start, RangeIndex):
+
+            if is_object_dtype(dtype):
+                return start.astype(object)
+            if not copy:
+                return start
+            if name is None:
+                name = getattr(start,'name',None)
+            start, stop, step = start.start, start.stop, start.step
+
+        # range
+        elif isinstance(start, range):
```

Comment:
Can we make a separate constructor for this? `RangeIndex.from_range`? Mostly this will just be called from `Index.__new__`.